num_enum_derive/src/lib.rs

```diff
@@ -145,6 +294,122 @@ pub fn derive_into_primitive(input: TokenStream) -> TokenStream {
+/// use num_enum::TryFromPrimitive;
+/// use std::convert::TryFrom;
+///
+/// #[derive(Debug, Eq, PartialEq, FromPrimitive)]
+/// #[repr(u8)]
+/// enum Number {
+///     Zero,
+///     #[num_enum(default)]
+///     NonZero,
+/// }
+///
+/// fn main() {
+///     let zero = Number::try_from(0u8);
+///     assert_eq!(zero, Ok(Number::Zero));
+///
+///     let one = Number::try_from(1u8);
+///     assert_eq!(one, Ok(Number::NonZero));
+///
+///     let two = Number::try_from(2u8);
+///     assert_eq!(two, Ok(Number::NonZero));
+/// }
+/// ```
+#[proc_macro_derive(FromPrimitive, attributes(num_enum))]
+pub fn derive_from_primitive(input: TokenStream) -> TokenStream {
+    let enum_info: EnumInfo = parse_macro_input!(input);
+    let krate = Ident::new(&get_crate_name(), Span::call_site());
+
+    // panic!("{:#?}", enum_info);
+
+    let CanonicalAndAlternatives {
+        canonical: canonical_idents,
+        alternatives: alternative_idents,
+    } = enum_info.idents();
+    let CanonicalAndAlternatives {
+        canonical: canonical_expressions,
+        alternatives: alternative_expressions,
+    } = enum_info.expressions();
+
+    debug_assert_eq!(canonical_idents.len(), canonical_expressions.len());
+    debug_assert_eq!(alternative_idents.len(), alternative_expressions.len());
+
+    let EnumInfo {
+        name,
+        repr,
+        default_variant,
+        ..
+    } = enum_info;
+
+    let default_ident = default_variant
+        .expect("#[derive(FromPrimitive)] requires a variant marked with `#[num_enum(default)]`");
+
+    TokenStream::from(quote! {
+        impl ::#krate::FromPrimitive for #name {
+            type Primitive = #repr;
+
+            fn from_primitive(number: Self::Primitive) -> Self {
+                #![deny(unreachable_patterns)]
+
+                // Use intermediate const(s) so that enums defined like
+                // `Two = ONE + 1u8` work properly.
+                #![allow(non_upper_case_globals)]
+                #(
+                    const #canonical_idents: #repr =
+                        #canonical_expressions
+                    ;
+                    #(
+                        const #alternative_idents: #repr =
+                            #alternative_expressions
+                        ;
+                    )*
+                )*
+                match number {
+                    #(
+                        | #canonical_idents #(| #alternative_idents )* => Self::#canonical_idents,
+                    )*
+                    | _ => Self::#default_ident,
+                }
+            }
+        }
+
+        impl ::core::convert::From<#repr> for #name {
+            #[inline]
+            fn from (
+                number: #repr,
+            ) -> Self {
+                ::#krate::FromPrimitive::from_primitive(number)
+            }
+        }
+
+        // The Rust stdlib will implement `#name: From<#repr>` for us for free!
+
+        impl ::#krate::TryFromPrimitive for #name {
```

Comment:
I'm still not completely convinced by this block... Can you explain the value it's bringing?

My reasoning goes: `TryFromPrimitive` only really exists as a trait so that the string representation of the enum name has somewhere static to live when implementing `std::convert::TryFrom` without needing to allocate - we don't expect anyone to call it directly. The compiler already provides a `std::convert::TryFrom` implementation with `Error=std::convert::Infallible` because of the `From` implementation block above this one. So I'm not sure when someone would be intentionally calling:
```rust
let e: Result<Number, num_enum::TryFromPrimitiveError<Number>> = Number::try_from_primitive(1);
```
instead of just
```rust
let e: Result<Number, std::convert::Infallible> = Number::try_from(1);
```

Do you have a use-case in mind where someone would want to explicitly use the `try_from_primitive` function instead? Or is it providing some other value I'm not seeing?

Reply:
The only reasoning was basically the "rule of least surprise" and trying to not make unexpected use-cases needlessly difficult/impossible: 

Given that `U: TryFrom<T>` is implied by `U: From<T>` I would expect `U: TryFromPrimitive<T>` to be implied by `U: FromPrimitive<T>`, too.

> `TryFromPrimitive` only really exists as a trait so that the string representation of the enum name has somewhere static to live when implementing `std::convert::TryFrom` without needing to allocate […]

Technically we wouldn't need `FromPrimitive<T>` then. Just deriving `From<T>` would work for the non-failing conversion.

But adding it gives us nice parallelism between safely and unsafely convertible enum types: both provide a type with a `Self::Primitive` type, that one might make us of. Skipping `FromPrimitive<T>` would make things asymmetric and might make it difficult to write generic functions that work with either.

In its current form the PR makes `FromPrimitive` a clean superset of `TryFromPrimitive`. Generically abstracting over types/traits works nicest with hierarchies of sub/supertypes.

| Trait                 | `#[derive(FromPrimitive)]` | `#[derive(FromPrimitive)]` | `#[derive(TryFromPrimitive)]` |
|                       | (Full)                     | (Minimal)                  |                               |
| --------------------- | -------------------------- | -------------------------- | ----------------------------- |
| `From<T>`             | ✅                          | ✅                          | ❌                             |
| `FromPrimitive<T>`    | ✅                          | ✅                          | ❌                             |
| `TryFrom<T>`          | ✅                          | ✅                          | ✅                             |
| `TryFromPrimitive<T>` | ✅                          | ❌                          | ✅                             |

Also you might want to make use of `Enum::Primitive` while in a fallible, generic context, which you could not do without.

This code wouldn't compile, e.g:

```rust
#[derive(FromPrimitive)]
#[repr(u8)]
enum SafeEnum {
    Zero = 0,
    #[num_enum(default)]
    NonZero = 1,
}

#[derive(TryFromPrimitive)]
#[repr(u8)]
enum UnsafeEnum {
    Zero = 0,
    One = 1,
}

fn from_u8<T>(byte: u8) -> Result<T, TryFromPrimitiveError<T>>
where
    T: TryFromPrimitive<Primitive = u8>
{
    byte.try_into()
}

let safe_enum: Result<SafeEnum, _> = from_u8(0x42);
let unsafe_enum: Result<UnsafeEnum, _> = from_u8(0x42);
```

By providing `Self::Primitive` the `TryFromPrimitive` trait provides useful additional information. Beyond merely being an implementation detail for `TryFrom`.

Reply:
> Given that `U: TryFrom<T>` is implied by `U: From<T>` I would expect `U: TryFromPrimitive<T>` to be implied by `U: FromPrimitive<T>`, too.
> ...
> Technically we wouldn't need `FromPrimitive<T>` then. Just deriving `From<T>` would work for the non-failing conversion.
> ...
> In its current form the PR makes `FromPrimitive` a clean superset of `TryFromPrimitive`. Generically abstracting over types/traits works nicest with hierarchies of sub/supertypes.

The thing is - there _is_ no `FromPrimitive` - it's not a trait which exists. If it existed, I think the symmetry would make more sense, but really this crate is about providing `From`, `TryFrom`, `Into`, and `TryInto` implementations, and `TryFromPrimitive` really only exists as an implementation detail, of the `TryFrom` implementation.

That said, `TryFromPrimitive` _does_ exist, and is necessarily part of the public API, so I think I'm sufficiently convinced by your example, even though I think it's generally better to avoid unnecessary fallibility where possible :)